### DISCUSSION TITLE
Hooks rateLimit applied / not applied

### DISCUSSION THREAD

**Discussion by mf-guse** [MarcGiffing/bucket4j-spring-boot-starter#347] [Ideas]
I would like to have hooks to register custom code which is executed when a rateLimit applied / did not apply.
It would be nice to register global hooks (maybe like metric handlers?) and lokal hooks.

Example global hook:

```
bucket4j:
   limited-hooks:
     -
   not-limited-hooks:
     - 
```

Example local hook:

```
    @RateLimiting(
         limitedHooks = {},
         notLimitedHooks = {}
    )
```

Is this a stupid idea? What would be the best way to implement this? Each hook is a SpEL expression? A Hook interface and several implementations?

**Comment by MarcGiffing** [comment]
For the local hooks the fallbackMethod can be used. Doesn't it suits  your use case? We can support an additional parameter with further information will will automatically be filled. 

```java
@RateLimiting(
            fallbackMethodName = "myFallbackMethod")
    public String execute(String username) {
        log.info("Method with Param {} executed", username);
        return username;
    }

    // the fallback method must have the same signature
    public String myFallbackMethod(String username, RateLimitInformation info) {
        log.info("Fallback-Method with Param {} executed with informations {}", username, info);

        return username;
    }
```

For the Servlets you can Implement the Interface MetricHandler and provide your custom code. What are you missing here?
https://github.com/MarcGiffing/bucket4j-spring-boot-starter/blob/master/examples/caffeine/src/main/java/com/giffing/bucket4j/spring/boot/starter/examples/caffeine/DebugMetricHandler.java

**Comment by mf-guse** [comment]
The idea was to do something when the rate limit applied / not applied like putting some information into a database, write a special log-entry or something like that.

Lets call it "cross-sectional-tasks", something which you man not want to implement every time again and again in the annotated method and the fallback method again, similar to the default metrics for filters.

**Comment by MarcGiffing** [comment]
That was the purpose of the MetricHandler. Why doesn't it suits your needs? Maybe we can use a better name?

**Comment by mf-guse** [comment]
Currently I'm quite unsure what my usecase is and if the MetricHandlers are enough. So I close this discussion to prevent confution (even on my side ;) ).
